codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -285,193 +326,506 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+        @Override
+        protected void doBeginRead() {
+            if (readInProgress) {
+                return;
+            }
+
+            final RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+            allocHandle.reset(config());
+            if (inboundBuffer.isEmpty()) {
+                readInProgress = true;
+                return;
+            }
+
+            do {
+                Object m = inboundBuffer.poll();
+                if (m == null) {
+                    break;
+                }
+                if (!doRead0(m, allocHandle)) {
+                    // Channel closed, and already cleaned up.
+                    return;
+                }
+            } while (allocHandle.continueReading());
+
+            allocHandle.readComplete();
+            pipeline().fireChannelReadComplete();
+        }
+
+        @Override
+        protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+            if (closed) {
+                throw CLOSED_CHANNEL_EXCEPTION;
+            }
+
+            try {
+                if (sizeEstimatorHandle == null) {
+                    sizeEstimatorHandle = config().getMessageSizeEstimator().newHandle();
+                }
+                for (;;) {
+                    final Object msg = in.current();
+                    if (msg == null) {
+                        break;
+                    }
+                    // TODO(buchgr): Detecting cancellation relies on ChannelOutboundBuffer internals. NOT COOL!
+                    if (msg == Unpooled.EMPTY_BUFFER /* The write was cancelled. */
+                        /* Write to trigger writability after window update. */
+                            || msg == REEVALUATE_WRITABILITY_MESSAGE) {
+                        in.remove();
+                        continue;
+                    }
+                    final int bytes = sizeEstimatorHandle.size(msg);
+                /*
+                 * The flow control window needs to be decrement before stealing the message from the buffer (and
+                 * thereby decrementing the number of pending bytes). Else, when calling steal() the number of pending
+                 * bytes could  be less than the writebuffer watermark (=flow control window) and thus trigger a
+                 * writability change.
+                 *
+                 * This code must never trigger a writability change. Only reading window updates or channel writes may
+                 * change the channel's writability.
+                 */
+                    incrementOutboundFlowControlWindow(-bytes);
+                    final ChannelPromise promise = in.steal();
+                    if (bytes > 0) {
+                        promise.addListener(new ReturnFlowControlWindowOnFailureListener(bytes));
+                    }
+                    // TODO(buchgr): Should we also the change the writability if END_STREAM is set?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)